### PR TITLE
Reword documentation for Service concept

### PR DESCRIPTION
**Pull request by sftim**
Reword [Service](https://kubernetes.io/docs/concepts/services-networking/service/) and the related [glossary entry](https://kubernetes.io/docs/reference/glossary/?all=true#term-service):

The _[Services v2](https://github.com/kubernetes/kubernetes/issues/1107)_ design document may be helpful when reviewing.

- Extend the “what's next” section
- Reword glossary entry for Service
  - New wording should be more suitable for inclusion via a `glossary_definition` shortcode.
- [Address the reader as “you”](https://kubernetes.io/docs/contribute/style/style-guide/#address-the-reader-as-you)
- Write Service in initial caps and not in backticks
- Adopt glossary definitions and tooltips where appropriate
- Wording improvements, eg switching to [active voice](https://kubernetes.io/docs/contribute/style/style-guide/#use-active-voice)
- Use [`note` shortcodes](https://kubernetes.io/docs/contribute/style/style-guide/#note) for notes
- Use [`warning` shortcodes](https://kubernetes.io/docs/contribute/style/style-guide/#warning) for warnings
- Use [documentation IP address range](https://tools.ietf.org/html/rfc5737) for examples

### Questions

- Should some of the `note` shortcodes actually be `caution` blocks instead?
- Is the new “motivation” section at the start of the page worded OK?
   - suggestions _very_ welcome
- Is this revised text OK from a technical accuracy standpoint?
- What have I missed?

### Changes set aside

The concept page about Service covers lots of topics. It has lots of details about exposing services using AWS cloud and about the internals of kube-proxy. Some of the page reads more like something that belongs in the Task section of docs.
To me those feel like good things to fix in one or more future PRs, after this lands. For example: a PR that migrates details of different operating modes for kube-proxy into the Reference section.

The diagrams in the page could use some care & maintenance. It's great to have them; they could be even greater.

I want to get this PR to the point where merging it represents a clear improvement over existing docs, get it merged, and leave room for future improvements (by me or anyone else).

## Connected issues
Relevant to #3735.

### CODE REVIEW

File: content/en/docs/concepts/services-networking/service.md [hunk 1]
```diff
@@ -14,44 +14,68 @@ weight: 10
...
+Each Pod gets its own IP address, however the set of Pods
+for a Deployment running in one moment in time could be different from
+the set of Pods running that application a moment later.
+
+This leads to a problem: if some set of Pods (call them “backends”) provides
+functionality to other Pods (call them “frontends”) inside your cluster,
+how do those frontends find out and keep track of which IP address to connect
+to, so that the frontend can use the backend part of the workload?
+
+Enter _Services_.
```
Comment (DanyC97) [inline content/en/docs/concepts/services-networking/service.md]: i think we can take this one one

File: content/en/docs/concepts/services-networking/service.md [hunk 2]
```diff
@@ -14,44 +14,68 @@ weight: 10
...
+
+In Kubernetes, a Service is an abstraction which defines a logical set of Pods
+and a policy by which to access them (you'll sometimes see this pattern called
+a micro-service). The set of Pods targeted by a Service is usually determined
+by a {{< glossary_tooltip text="selector" term_id="selector" >}}
+(see [below](#services-without-selectors) for why you might want a Service
+_without_ a selector).
+
+For example: consider a stateless image-processing backend which is running with
+3 replicas.  Those replicas are fungible&mdash;frontends do not care which backend
```
Comment (DanyC97) [inline content/en/docs/concepts/services-networking/service.md]: nit: double space - up to you if you want to fix it

File: content/en/docs/concepts/services-networking/service.md [hunk 3]
```diff
@@ -14,44 +14,68 @@ weight: 10
...
+
+The Service abstraction enables this decoupling.
+
+### Cloud-native service discovery
+
+If you're able to use Kubernetes APIs for service discovery in your application,
+you can query the {{< glossary_tooltip text="API server" term_id="kube-apiserver" >}}
+for Endpoints, that will be updated whenever the set of Pods in a Service changes.
+
+For non-native applications, Kubernetes offers ways to place a network port or load
```
Comment (DanyC97) [inline content/en/docs/concepts/services-networking/service.md]: nit: replace `balancer to between` -> `balancer in between`

File: content/en/docs/concepts/services-networking/service.md [hunk 4]
```diff
@@ -14,44 +14,68 @@ weight: 10
...
+balancer to between your application and the backend Pods.
+
 ## Defining a service
 
-A `Service` in Kubernetes is a REST object, similar to a `Pod`.  Like all of the
-REST objects, a `Service` definition can be POSTed to the apiserver to create a
-new instance.  For example, suppose you have a set of `Pods` that each expose
-port 9376 and carry a label `"app=MyApp"`.
+A Service in Kubernetes is a REST object, similar to a Pod.  Like all of the
```
Comment (DanyC97) [inline content/en/docs/concepts/services-networking/service.md]: pls remove `can be`

File: content/en/docs/concepts/services-networking/service.md [hunk 5]
```diff
@@ -67,40 +91,52 @@ spec:
...
+version of your backend software, without breaking clients.
+
+The default protocol for services is TCP; you can also use any other
+[supported protocol](#protocol-support).
+
+As many Services need to expose more than one port, Kubernetes supports multiple
+port definitions on a Service object.
+Each port definition can have the same `protocol`, or a different one.
 
 ### Services without selectors
```
Comment (DanyC97) [inline content/en/docs/concepts/services-networking/service.md]: i'd put serverless in the brackets as folks are used with this _naming_ too

File: content/en/docs/concepts/services-networking/service.md [hunk 6]
```diff
@@ -124,23 +161,25 @@ metadata:
...
     ports:
       - port: 9376
 ```
 
 {{< note >}}
-The endpoint IPs may not be loopback (127.0.0.0/8), link-local
-(169.254.0.0/16), or link-local multicast (224.0.0.0/24). They cannot be the
-cluster IPs of other Kubernetes services either because the `kube-proxy`
-component doesn't support virtual IPs as destination yet.
+The endpoint IPs _may not_ be: loopback (127.0.0.0/8 for IPv4, ::1/128 for IPv6), or
```
Comment (DanyC97) [inline content/en/docs/concepts/services-networking/service.md]: you think it makes sense to s/may/cannot ?

File: content/en/docs/concepts/services-networking/service.md [hunk 7]
```diff
@@ -150,90 +189,129 @@ Every node in a Kubernetes cluster runs a `kube-proxy`. `kube-proxy` is
...
 than [`ExternalName`](#externalname).
 
-In Kubernetes v1.0, `Services` are a "layer 4" (TCP/UDP over IP) construct, the
-proxy was purely in userspace.  In Kubernetes v1.1, the `Ingress` API was added
-(beta) to represent "layer 7"(HTTP) services, iptables proxy was added too,
-and became the default operating mode since Kubernetes v1.2. In Kubernetes v1.8.0-beta.0,
-ipvs proxy was added.
+### Why not use round-robin DNS?
+
+A question that pops up every now and then is why Kubernetes has uses proxying
```
Comment (DanyC97) [inline content/en/docs/concepts/services-networking/service.md]: i'd reword `why Kubernetes has uses proxying to direct incoming traffic to backends` to

`why Kubernetes need to proxy the incoming direct traffic to backends`

File: content/en/docs/concepts/services-networking/service.md [hunk 8]
```diff
@@ -150,90 +189,129 @@ Every node in a Kubernetes cluster runs a `kube-proxy`. `kube-proxy` is
...
+   caching the results of name lookups.
+ * Many apps do DNS lookups only once and cache the results indefinitely.
+ * Even if apps and libraries did proper re-resolution, the load of every
+   client re-resolving DNS over and over would be difficult to manage.
+
+### Version compatibility
+
+Services are a "layer 4" (TCP/UDP over IP) construct. Since Kubernetes v1.0
+you have been able to use the [userspace proxy mode](#proxy-mode-userspace).
+Kubernetes v1.1 added iptables mode proxying was added too, which became the
```
Comment (DanyC97) [inline content/en/docs/concepts/services-networking/service.md]: i'd reword `Kubernetes v1.1 added iptables mode proxying was added too, which became the` to

`Kubernetes v1.1 added iptables proxy mode which became the`

File: content/en/docs/concepts/services-networking/service.md [hunk 9]
```diff
@@ -150,90 +189,129 @@ Every node in a Kubernetes cluster runs a `kube-proxy`. `kube-proxy` is
...
+removal of Service and Endpoint objects. For each Service it opens a
 port (randomly chosen) on the local node.  Any connections to this "proxy port"
-will be proxied to one of the `Service`'s backend `Pods` (as reported in
-`Endpoints`).  Which backend `Pod`  to use is decided based on the
-`SessionAffinity` of the `Service`.  Lastly, it installs iptables rules which
-capture traffic to the `Service`'s `clusterIP` (which is virtual) and `Port`
-and redirects that traffic to the proxy port which proxies the backend `Pod`.
-By default, the choice of backend is round robin.
+will be proxied to one of the Service's backend Pods (as reported via
+Endpoints).  Which backend Pod  to use is decided based on the
```
Comment (DanyC97) [inline content/en/docs/concepts/services-networking/service.md]: i'd replace `Which pod` with `The pod to use is ...`

File: content/en/docs/concepts/services-networking/service.md [hunk 10]
```diff
@@ -150,90 +189,129 @@ Every node in a Kubernetes cluster runs a `kube-proxy`. `kube-proxy` is
...
+`SessionAffinity` of the Service.
+
+Lastly, the user-space proxy installs iptables rules which capture traffic to
+the Service's `clusterIP` (which is virtual) and `port`. The rules
+redirect that traffic to the proxy port which proxies the backend Pod.
+
+By default, kube-proxy in userspace mode chooses a backend via a round-robin algorithm.
```
Comment (DanyC97) [inline content/en/docs/concepts/services-networking/service.md]: nit: will be nice if we had some info about the round-robin algorithm so we can link it (i know this is for advanced users but some developers for examples will appreciate the details

File: content/en/docs/concepts/services-networking/service.md [hunk 11]
```diff
@@ -150,90 +189,129 @@ Every node in a Kubernetes cluster runs a `kube-proxy`. `kube-proxy` is
...
+ ipvs rules with Kubernetes Services and Endpoints periodically.
+This control loop ensures that ipvs status matches the desired
+state.
+When accessing a Service, ipvs will direct traffic to one of the backend Pods.
+
+The ipvs proxy mode is based on netfilter hook function that is similar to
+iptables mode, but uses hash table as the underlying data structure and works
+in the kernel space.
+That means kube-proxy in ipvs mode redirects traffic with a lower latency than
+kube-proxy in iptables mode, wih much better performance when synchronising
```
Comment (DanyC97) [inline content/en/docs/concepts/services-networking/service.md]: typo `wih` -> `with`

File: content/en/docs/concepts/services-networking/service.md [hunk 12]
```diff
@@ -150,90 +189,129 @@ Every node in a Kubernetes cluster runs a `kube-proxy`. `kube-proxy` is
...
+proxy rules. ipvs mode also supports a higher throughput of network traffic.
```
Comment (DanyC97) [inline content/en/docs/concepts/services-networking/service.md]: should we add smicolon -> `proxy rules; ipvs mode also supports a`

File: content/en/docs/concepts/services-networking/service.md [hunk 13]
```diff
@@ -150,90 +189,129 @@ Every node in a Kubernetes cluster runs a `kube-proxy`. `kube-proxy` is
...
 
-Similar to iptables, Ipvs is based on netfilter hook function, but uses hash
-table as the underlying data structure and works in the kernel space.
-That means ipvs redirects traffic much faster, and has much
-better performance when syncing proxy rules. Furthermore, ipvs provides more
-options for load balancing algorithm, such as:
+ipvs provides more options for balancing traffic to backend Pods; these are:
```
Comment (DanyC97) [inline content/en/docs/concepts/services-networking/service.md]: capitalize ipvs ?

File: content/en/docs/concepts/services-networking/service.md (hunk 1 above)
Comment (sftim) [reply content/en/docs/concepts/services-networking/service.md]: ?

File: content/en/docs/concepts/services-networking/service.md (hunk 3 above)
Comment (sftim) [reply content/en/docs/concepts/services-networking/service.md]: Thanks for spotting.

File: content/en/docs/concepts/services-networking/service.md (hunk 6 above)
Comment (sftim) [reply content/en/docs/concepts/services-networking/service.md]: I'll try “must not”

File: content/en/docs/concepts/services-networking/service.md (hunk 10 above)
Comment (sftim) [reply content/en/docs/concepts/services-networking/service.md]: This PR is already huge! That feels like something to fix in a follow-up set of changes.

File: content/en/docs/concepts/services-networking/service.md (hunk 12 above)
Comment (sftim) [reply content/en/docs/concepts/services-networking/service.md]: I'll reword. Aim here is to write in a way that lends itself to localization.

File: content/en/docs/concepts/services-networking/service.md (hunk 6 above)
Comment (sftim) [reply content/en/docs/concepts/services-networking/service.md]: How's that?

File: content/en/docs/concepts/services-networking/service.md (hunk 7 above)
Comment (sftim) [reply content/en/docs/concepts/services-networking/service.md]: I reworded this.

File: content/en/docs/concepts/services-networking/service.md (hunk 8 above)
Comment (sftim) [reply content/en/docs/concepts/services-networking/service.md]: Reworded

File: content/en/docs/concepts/services-networking/service.md (hunk 1 above)
Comment (DanyC97) [reply content/en/docs/concepts/services-networking/service.md]: sorry, had a typo. Personally i don't think we need the text above

File: content/en/docs/concepts/services-networking/service.md (hunk 6 above)
Comment (DanyC97) [reply content/en/docs/concepts/services-networking/service.md]: sounds good to me

File: content/en/docs/concepts/services-networking/service.md (hunk 10 above)
Comment (DanyC97) [reply content/en/docs/concepts/services-networking/service.md]: okay, fair point

File: content/en/docs/concepts/services-networking/service.md (hunk 1 above)
Comment (sftim) [reply content/en/docs/concepts/services-networking/service.md]: @DanyC97 do you favour removing or rewording?

Comment (DanyC97) [reply content/en/docs/concepts/services-networking/service.md]: @sftim i was in favour of removing it however i see folks are using it in blog posts so no longer a strong view on it ;)

**Comment by DanyC97** [comment]
great piece of work @sftim thanks for taking into account my feedback.
/lgtm

**Comment by DanyC97** [comment]
@zhangxiaoyu-zidif please help review and approve, thx

File: content/en/docs/concepts/services-networking/service.md [hunk 14]
```diff
@@ -310,43 +384,53 @@ REDIS_MASTER_PORT_6379_TCP_PORT=6379
...
-cluster IP (`.spec.clusterIP`).
+this case, you can create what are termed “headless” Services, by explicitly
+specifying `"None"` for the cluster IP (`.spec.clusterIP`).
 
-This option allows developers to reduce coupling to the Kubernetes system by
-allowing them freedom to do discovery their own way.  Applications can still use
-a self-registration pattern and adapters for other discovery systems could easily
+You can use a headless Service to interface with other service discovery mechanisms,
+without being tied to Kubernetes' implementation. For example, you could implement
+a custom [Operator](
```
Comment (lalyos) [inline content/en/docs/concepts/services-networking/service.md]: This line seems to be unfinished. @sftim could you please verify?

Comment (sftim) [reply content/en/docs/concepts/services-networking/service.md]: Now fixed.

File: content/en/docs/concepts/services-networking/service.md [hunk 15]
```diff
@@ -394,36 +477,47 @@ The default is `ClusterIP`.
      load balancer will route, are automatically created.
    * [`ExternalName`](#externalname): Maps the service to the contents of the
      `externalName` field (e.g. `foo.bar.example.com`), by returning a `CNAME` record
-     with its value. No proxying of any kind is set up. This requires version 1.7 or
-     higher of `kube-dns`.
+     with its value. No proxying of any kind is set up.
+
+{{< note >}}
+
+You need CoreDNS version 1.7 or higher to use the `ExternalName` type.
```
Comment (wozniakjan) [inline content/en/docs/concepts/services-networking/service.md]: this swapped CoreDNS and kube-dns without changing the versioning

Comment (sftim) [reply content/en/docs/concepts/services-networking/service.md]: Oops!

@wozniakjan would you be willing to [log an issue](https://github.com/kubernetes/website/issues/new?labels%5B%5D=kind%2Fbug&template=bug-report.md) if appropriate?

Comment (wozniakjan) [reply content/en/docs/concepts/services-networking/service.md]: I skipped ahead and went for a PR https://github.com/kubernetes/website/pull/18566 if that is ok

Comment (sftim) [reply content/en/docs/concepts/services-networking/service.md]: Yeah, that's great!
